xarray/core/dataarray.py

```diff
@@ -484,7 +484,13 @@ def __setitem__(self, key, value):
         if isinstance(key, basestring):
             self.coords[key] = value
         else:
-            # xarray-style array indexing
+            # Coordinates in key, value and self[key] should be consistent.
+            obj = self[key]
+            if isinstance(value, DataArray):
+                assert_coordinate_consistent(value, obj.coords.variables)
+            # DataArray key -> Variable key
+            key = {k: v.variable if isinstance(v, DataArray) else v
+                   for k, v in self._item_key_to_dict(key).items()}
```

Comment:
Do we need to check that coordinates are consistent on the `key`?

Reply:
This is done a few lines above, `obj = self[key]`, where in `.isel` we check the coordinates in the `key`.

But I am wondering this unnecessary indexing, though I think this implementation is the simplest.

Reply:
Hmm. This could indeed be a significant performance hit. That said, I'm OK leaving this for now, with a TODO note to optimize it later.